libraries/state_history/log.cpp

```diff
@@ -288,21 +313,88 @@ void state_history_log::split_log() {
    index.open("w+b");
 }
 
+void state_history_log::store_entry(const chain::block_id_type& id, const chain::block_id_type& prev_id,
+                                    std::vector<char>&& data) {
+   
+   if (thr.joinable()) {
+      if (write_thread_has_exception) {
+         std::rethrow_exception(eptr);
+      }
+      block_num_type block_num = chain::block_header::num_from_id(id);
+      auto shared_data = std::make_shared<std::vector<char>>(std::move(data));
+      boost::asio::post(work_strand, [this, entry = log_entry_type{id, prev_id, shared_data} ]() {
+         write_entry(entry);
+      });
+      cached[block_num] = shared_data;
+
+      while (cached.size() > num_buffered_entries && cached.begin()->second.use_count() == 1) {
+         // the log entry has been removed from write_queue, so we can safely erase it. 
+         cached.erase(cached.begin());
+      }
+
+      fc_ilog(logger,"store_entry name=${name}, block_num=${block_num} cached.size = ${sz}, num_buffered_entries=${num_buffered_entries}, id=${id}",
+            ("name", name)("block_num", block_num)("sz", cached.size())("num_buffered_entries", num_buffered_entries)("id", id));
+
+   }
+   else {
```

Comment:
`else` is no longer needed, correct? The thread should always be joinable and writes will always happen in other thread.